[Done] Translation of tutorial.md .. (#135)
reactjs/ar.reactjs.org: content/tutorial/tutorial.md

---
id: tutorial

> title: الدليل التطبيقي: مدخل إلى React

Reviewer: This line makes the build fail, make it between `" "`
```suggestion
title: "الدليل التطبيقي: مدخل إلى React"
```

  - "docs/tutorial-zh-CN.html"
---

لا يفترض هذا الدليل أي معرفة مسبقة بمكتبة React.

> ## قبل أن نبدأ بالدليل التطبيقي {#before-we-start-the-tutorial}

Reviewer 2: قبل أن نبدأ بالدورة التعليمية> قبل أن نبدأ بالدليل التطبيقي

سنبني لعبة صغيرة خلال هذا الدليل التطبيقي. **ربّما قد ترغب بتخطي هذا الدليل لأنّك لا تريد بناء الألعاب، ولكن أعطيها فرصة.** إنّ التقنيات التي ستتعلمها في هذا الدليل أساسيّة لبناء أي تطبيق React، وسيعطيك إتقانها فهمًا أعمق لمكتبة React.

> >فائدة

Reviewer 2: فائدة > نصائح

>

> >هذا الدليل مُصمَّم للأشخاص الذين يُفضّلون **التعلّم بالممارسة**. إن كنت تُفضّل تعلّم المفاهيم من البداية فارجع إلى [مستندات React من البداية خطوة بخطوة](/docs/hello-world.html). قد تجد هذا الدليل متكاملًا مع مستندات React.

Reviewer: Other suggestions:
1. Keep it as it is :smile:
2. or:
```suggestion
>هذا الدليل مُصمَّم للأشخاص الذين يُفضّلون **التعلّم بالممارسة**. إن كنت تُفضّل تعلّم المفاهيم من البداية انتقل إلى [مستندات React من البداية خطوة بخطوة](/docs/hello-world.html). قد تجد هذا الدليل متكاملًا مع مستندات React.
```

3. or:
```suggestion
>هذا الدليل مُصمَّم للأشخاص الذين يُفضّلون **التعلّم بالممارسة**. إن كنت تُفضّل تعلّم المفاهيم من البداية فانتقل إلى [مستندات React من البداية خطوة بخطوة](/docs/hello-world.html). قد تجد هذا الدليل متكاملًا مع مستندات React.
```

4. or:
```suggestion
>هذا الدليل مُصمَّم للأشخاص الذين يُفضّلون **التعلّم بالممارسة**. إن كنت تُفضّل تعلّم المفاهيم من البداية عُد إلى [مستندات React من البداية خطوة بخطوة](/docs/hello-world.html). قد تجد هذا الدليل متكاملًا مع مستندات React.
```

@mohshbool, what do you think?

Author: 5. or:
change ` فارجع إلى` to `راجع` 😄

Reviewer: طبعًا :smile: !

> هذا الدليل مقسوم إلى عدّة أقسام:

Reviewer 2: هذه الدورة مقسمة إلى عدّة أقسام:

* يُعطيك قسم [الإعداد من أجل الدليل](#setup-for-the-tutorial) **نقطة بداية** لمتابعة الدليل.
* يُعلّمك قسم [لمحة عامّة](#overview) **أساسيات** React: المكوّنات، والخاصيّات، والحالة.

> * يُعلِّمك قسم [إكمال اللعبة](#completing-the-game) **أشيع التقنيات** في تطوير React.

Author: لا داعي لتبيسط كل الكلمات .. يجب ان يكون هنك إثراء لغوي في المصطلحات المستعملة

> * يُعلِّمك قسم [إكمال اللعبة](#completing-the-game) **أشيع التقنيات** في تطوير React.

Reviewer: <div dir="rtl">
ليست بغرض التبسيط فحسب وإنما لتحسين التعبير، فاستعمال كلمة "أَشْيَعْ" ليس واردا كثيرا (عن نفسي لم أصادفه من قبل). كلمة أشْيَع كذلك (وإن أستعملت، وَجَب تشكيلها لوجود كلمة "أُشِيع" والتي تَرد أكثر في التقارير والكُتب...
</div>

> * يُعلِّمك قسم [إكمال اللعبة](#completing-the-game) **أشيع التقنيات** في تطوير React.

Author: اظن انها كلمة متعارف عليها في اللغة ... لكن لاباس ساقوم بتغيرها ✌️ ..

### المتطلّبات الأساسيّة {#prerequisites}

سنفترض أنّك متآلف مع HTML و JavaScript، ولكن يجب أن تكون قادرًا على المتابعة حتى ولو كنت قادمًا من لغة برمجة أخرى. سنفترض أنّك متآلف مع المفاهيم البرمجيّة مثل الدوال، والكائنات، والمصفوفات، وبدرجة أقل الأصناف.

إن احتجت لمراجعة JavaScript نوصيك بالرجوع إلى [مستندات JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript) في موسوعة حسوب. لاحظ أنّنا نستخدم بعض الميزات من ES6، وهي إصدار جديد من JavaScript. سنستخدم في هذا الدليل [الدوال السهمية](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)، [الأصناف](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)، والتصريحين [`let`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) و [`const`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const). بإمكانك استخدام [Babel REPL](babel://es5-syntax-example) لتتحقّق إلى ماذا تُصرَّف شيفرة ES6.

## الإعداد من أجل الدليل {#setup-for-the-tutorial}

هناك طريقتان لإكمال هذا الدليل، بإمكانك إمّا كتابة الشيفرة في متصفحك، أو إعداد بيئة تطوير محلية على حاسوبك.

> ### الخيار الأول للإعداد: كتابة الشيفرة في المتصفح {#setup-option-1-write-code-in-the-browser}

Reviewer 2: الإعدادات الخيار 1 > الخيار الأول للإعداد

هذه أسرع طريقة للبدء.

في البداية افتح **[هذه الشيفرة المبدئيّة](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)** في نافذة جديدة. يجب أن تعرض النافذة الجديدة لوحة لعبة إكس-أو وشيفرة React. سنُعدِّل شيفرة React في هذا الدليل.

بإمكانك الآن تجاوز الخيار الثاني للإعداد والذهاب إلى قسم [لمحة عامّة](#overview) للحصول على لمحة عامّة عن React.

> ### الخيار الثاني للإعداد: بيئة التطوير المحليّة {#setup-option-2-local-development-environment}

Reviewer 2: الخيار الثاني للإعداد > الإعدادات الخيار 2

}
```

> قبل التغييرات:

Reviewer 2: قبل> قبل التغييرات

Author: Here just for clarification ..

![React Devtools](../images/tutorial/tictac-empty.png)

> بعد التغييرات: يجب أن ترى عددًا في كل مربّع من الناتج المُصيَّر:

Reviewer 2: بعد التغييرات: يجب أن ترى عددًا في كل مربّع من الناتج المُصيَّر:
 > بعد : يجب أن ترى عددًا في كل مربّع من الناتج المُصيَّر:

Author: Here just for clarification too ..

var newPlayer = Object.assign({}, player, {score: 2});
// الآن يبقى player بدون تغيير
// ولكن تكون قيمة newPlayer هي {score: 2, name: 'Jeff'}

/// أو إن كنت تستخدم اقتراح صياغة نشر الكائنات تستطيع كتابة
// var newPlayer = {...player, score: 2};
```

النتيجة النهائيّة هي نفسها ولكن نكسب عدّة فوائد عن طريق عدم تغيير البيانات بشكل مباشر، سنذكر هذه الفوائد الآن.

#### تبسيط الميزات المعقدة {#complex-features-become-simple}

تجعل عدم قابلية التغيير من الأسهل تنفيذ الميزات المعقدة. لاحقًا في هذا الدليل سنُنفِّذ ميزة السفر عبر الزمن والتي تسمح لنا بمراجعة تاريخ الحركات السابقة في لعبة إكس-أو مع إمكانية القفز إلى الحركات السابقة. هذه الميزة ليست خاصة بالألعاب، فالقدرة على التراجع والعودة عن أفعال محددة هي متطلب شائع في التطبيقات. يسمح لك تجنّب التعديل المباشر للبيانات بالاحتفاظ بإصدارات من تاريخ تحركات اللعبة وإعادة استخدامها لاحقًا.

#### كشف التغيرات {#detecting-changes}

يصعب كشف التغيّرات في الكائنات القابلة للتعديل لأنّها تُعدَّل بشكل مباشر. يتطلب هذا الكشف مقارنة الكائنات القابلة للتعديل مع النسخ السابقة منها وكامل شجرة الكائنات المطلوبة.

يُعتبَر كشف التغيّرات في الكائنات غير القابلة للتعديل أسهل. إن كان الكائن غير القابل للتعديل مختلفًا عن السابق فنعتبر أنّ الكائن قد تغيّر.

> #### تحديد وقت إعادة التصيير في React {#determining-when-to-re-render-in-react}

Reviewer: For me, the "when to re-render" expression is related to time=>timers which is not the case for us. I believe that using `متى` is better since it refers to "reaction to an event".

```suggestion
#### تحديد متى يتم إعادة التصيير في React {#determining-when-to-re-render-in-react}
```

Or we should find a better expression

Author: `تحديد زمن  إعادة التصيير في React` !!

Reviewer: <div dir="rtl">
نفس الشيء، إعادة التصيير ليست متعلقة بالزمن/وقت وإنما بالأحداث على العموم، فهل يتم ذلك بإحدى توابع المؤقتات أم بواسطة shouldComponentUpdate.

نقول، "I sleep when I get tired" بمعنى "أنام حين أتعب" أو "أنام متى شعرت بالتعب" و عليه موعد النوم في هذا المثال مُحدد بحدث وليس زمن مُعيّن.
</div>

Author: لا تجوز المقارنة اللغوية بين اللغة الإنجليزية و العربية .. و اظن انك تعلم هذا جيدا ... 

برغم من عدم إقتناعي الشخصي بتغير بعض الكلمات لكن لاباس بهذا تحب إطار اوبن مايند 🤖 👽 ههه ..

  }
```

بها التغيير تستطيع `X` و `O` أخذ الأدوار. 

فلنُغيِّر أيضًا نص الحالة في التابع `render` للمكوّن `Board` بحيث يعرض من هو اللاعب الذي سيلعب الدور التالي:

```javascript{2}
  render() {

>     const status = 'اللاعب التالي: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: This could be confusing since we didn't translate the screenshots from the game nor the inspect-element screens nor the codepen examples.

Our convention is to translate the code examples, but I think that we can drop this rule for now, and we comeback to this section by translating screenshots & codepens... So we just keep it like this for now and refere that "Next player" means "اللاعب التالي"
```suggestion
    const status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

  }

  render() {

>     const status = 'اللاعب التالي: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: Same
```suggestion
    const status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

}
```

سنستدعي التابع `calculateWinner(squares)`‎ في تابع التصيير `render` للمكوّن `Board` للتحقّق من فوز اللاعب. إن فاز اللاعب فنستطيع عرض نص مثل "الفائز: X" أو "الفائز: O". سنضع هذه الشيفرة بدلًا من تصريح `الحالة` الموجود في التابع `render` للمكوّن `Board`:

```javascript{2-8}
  render() {
    const winner = calculateWinner(this.state.squares);
    let status;
    if (winner) {
      status = 'الفائز: ' + winner;
    } else {

>       status = 'اللاعب التالي: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ٍSame
```suggestion
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

    const winner = calculateWinner(this.state.squares);
    let status;
    if (winner) {
      status = 'الفائز: ' + winner;
    } else {

>       status = 'اللاعب التالي: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

    let status;
    if (winner) {
      status = 'الفائز: ' + winner;
    } else {

>       status = 'اللاعب التالي: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

    let status;
    if (winner) {
      status = 'الفائز: ' + winner;
    } else {

>       status = 'اللاعب التالي: ' + (this.state.xIsNext ? 'X' : 'O');

Reviewer: ```suggestion
      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
```

  }
```

**[انظر إلى كامل الشيفرة عند هذه النقطة](https://codepen.io/gaearon/pen/EmmGEa?editors=0010)**

نُنشِئ لكل تحرك في تاريخ اللعبة عنصرًا بالقائمة `<li>` يحتوي على زر `<button>`. يمتلك الزر مُعالِج للأحداث `onClick` والذي يستدعي تابًعا يُدعى `this.jumpTo()`‎. لم ننفذ هذا التابع حتى الآن. ينبغي أن نرى الآن قائمة بالتحركات التي حصلت في اللعبة وتحذير في النافذة console ضمن أدوات التطوير يقول:

>  تحذير:
>  يجب أن يمتلك كل عنصر ابن في المصفوفة خاصية مفتاح `key`. تحقق من التابع `render` للمكوّن `Game`.

فلنناقش معنى التحذير السابق.

### انتقاء مفتاح {#picking-a-key}

عند تصيير قائمة تُخزن React بعض المعلومات حول كل عنصر منها. عندما نحدث القائمة تحتاج React لتحديد ما تغيّر. بإمكاننا إضافة، وإزالة، وإعادة ترتيب، وتحديث عناصر القائمة.

تخيل الانتقال من:

```html
<li>أليكسا: 7 مهام متبقية</li>

> <li>بن: 5 مهام متبقية</li>

Reviewer: I suggest renaming `Ben` to any name, even if it was an Arabic name or at least use tashkeel `التشكيل` since it will be "kind of" confusing later when we elaborate on:

```
In addition to the updated counts, a human reading this would probably say that we swapped Alexa and Ben's ordering and inserted Claudia between Alexa and Ben. However, React is a computer program and does not know what we intended. Because React cannot know our intentions, we need to specify a *key* property for each list item to differentiate each list item from its siblings. One option would be to use the strings `alexa`, `ben`, `claudia`. If we were displaying data from a database, Alexa, Ben, and Claudia's database IDs could be used as keys.
```

```suggestion
<li>بَنْ: 5 مهام متبقية</li>
```

```

إلى:

```html

> <li>بن: 9 مهام متبقية</li>

Reviewer: ```suggestion
<li>بَنْ: 9 مهام متبقية</li>
```

<li>كلوديا: 8 مهام متبقية</li>
<li>أليكسا: 5 مهام متبقية</li>
```

> من وجهة نظرنا، بدل الانتقال بين ترتيب أليكسا وبن وأدخل كلوديا بينهما، ولكنّ React هي برنامج حاسوب ولا تفهم مقصدنا من ذلك، نحتاج لتحديد خاصيّة مفتاح *key* لكل عنصر قائمة لتمييزه عن العناصر الأشقاء له. بإمكاننا استخدام السلاسل النصيّة "`أليكسا`"، و "`بن`"، و "`كلوديا`" كمفاتيح. إن كان لدينا الوصول إلى قاعدة البيانات فنستطيع استخدام المُعرِّفات (IDs) لأليكسا وكلوديا وبن من قاعدة البيانات كمفاتيح:

Reviewer: ```suggestion
من وجهة نظرنا، بدل الانتقال بين ترتيب أليكسا وبن وأدخل كلوديا بينهما، ولكنّ React هي برنامج حاسوب ولا تفهم مقصدنا من ذلك، نحتاج لتحديد خاصيّة مفتاح *key* لكل عنصر قائمة لتمييزه عن العناصر الأشقاء له. بإمكاننا استخدام السلاسل النصيّة "`أليكسا`"، و "`بَنْ`"، و "`كلوديا`" كمفاتيح. إن كان لدينا الوصول إلى قاعدة البيانات فنستطيع استخدام المُعرِّفات (IDs) لأليكسا وكلوديا وبَنْ من قاعدة البيانات كمفاتيح:
```

    });
```

**[انظر إلى كامل الشيفرة عند هذه النقطة](https://codepen.io/gaearon/pen/PmmXRE?editors=0010)**

> يرمي النقر على أي زر من عناصر القائمة خطأً لأنّ التابع `jumpTo` غير مُعرَّف. قبل تنفيذ التابع `jumpTo` سنُضيف رقم الخطوة `stepNumber` إلى حالة مكوّن اللعبة `Game` للإشارة إلى رقم الخطوة التي نشاهدها حاليًّا.

Reviewer 3: ```suggestion
سيُظهر "يرمي" النقر على أي زر من عناصر القائمة خطأً لأنّ التابع `jumpTo` غير مُعرَّف. قبل تنفيذ التابع `jumpTo` سنُضيف رقم الخطوة `stepNumber` إلى حالة مكوّن اللعبة `Game` للإشارة إلى رقم الخطوة التي نشاهدها حاليًّا.
```

Author: أظن ان فعل `يرمي` هنا واضح الدلالة و المعنى خاصة في هذا السياق ..

Reviewer: أرى أن محمد مُحقّ في اقتراحه، ففي كثير من التعريبات تترجم عبارة "يرمي الخطأ" إلى "يُظهر الخطأ" أو "يُظهر رسالة الخطأ" واقتراح محمّد هو الوسط، حيث نترك التعريب التقني المعتاد وكذا التعريب الحرفي لهؤلاء المُعتادين على كلتا العبارتين "المُعرّبة والأصلية بالإنجليزية".
فكما نعلم، هناك من يعتمد على المصادر العربية او التعريب الآلي بشكل شبه كامل وبالتالي بعض العبارات التي اعتدناها قد تبدو غريبة لهم

Author: كلامك منطقي .. فقد تبدو غريبة لمن هجر العربية و يظن انه عربي ... و للاسف هذا حال الأمة اليوم

  }
```

أخيرًا سنُعدِّل التابع `render` للمكوّن `Game` من تصيير التحرك الأخير دومًا إلى تصيير التحرك المُختار حاليًّا وفق قيمة `stepNumber`:

```javascript{3}
  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);

    // بقية الشيفرة لم تتغير
```

إن نقرنا على أي تحرك في تاريخ اللعبة، فستُحدَّث لوحة اللعبة فورًا لتظهر بالشكل المفروض بعد حدوث ذلك التحرّك.

**[انظر إلى كامل الشيفرة عند هذه النقطة.](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**

### الملخص {#wrapping-up}

> تهانينا! لقد أنشأتَ لعبة إكس-أو تفعل ما يلي:

Reviewer 3: ```suggestion
تهانينا! لقد أنشأتَ لعبة إكس-أو والَّتي:
```

* تُتيح لك لعب إكس-أو.
* تُخبرنا عندما يفوز اللاعب.
* تُخزِّن تاريخ اللعبة أثناء تقدمّ اللعبة.
* تسمح للاعبين بمراجعة تاريخ اللعبة ورؤية الإصدارات القديمة للوحة اللعبة.

عمل رائع! نتمنى أن تشعر الآن بأنك تمتلك لمحة حول آليّة عمل React.

تحقّق من النتيجة النهائيّة هن : **[النتيجةالنهائيّة](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**.

إن كان لديك المزيد من الوقت أو أردت ممارسة مهاراتك الجديدة التي تعلمتها في React، فهذه بعض الأفكار للتحسين والتي قد تجريها على لعبة إكس-أو وهي مرتبة بشكل تصاعدي حسب الصعوبة:

1. عرض موقع كل تحرك بصيغة العامود والصف في قائمة تاريخ التحركات.
2. تعيين تنسيق عريض (Bold) للعنصر المُحدَّد حاليًّا في قائمة التحركات.
3. إعادة كتابة المكوّن `Board` ليستخدم حلقتين لإنشاء المربعات بدلًا من كتابتها حرفيًّا.
4. إضافة زر لترتيب التحركات تصاعديًّا أو تنازليًّا.
5. تعليم المربعات الثلاث عند فوز أحد اللاعبين والتي سببت فوزه.
6. عرض رسالة بنتيجة التعادل عندما لا يفوز أحد.

> تعلمنا خلال هذا الدليل مفاهيم في React تتضمّن العناصر، والمكوّنات، والخاصيّات، والحالة. للحصول على شرح مُفصَّل عن كل من هذه المواضيع تحقّق من [باقي مستندات React](/docs/hello-world.html). لتعلّم المزيد حول تعريف المكوّنات تحقّق من مرجع [الصنف `React.Component`](/docs/react-component.html).

Reviewer 3: is there a reason behind remove the `API` keyword?
```suggestion
تعلمنا خلال هذا الدليل مفاهيم في React تتضمّن العناصر، والمكوّنات، والخاصيّات، والحالة. للحصول على شرح مُفصَّل عن كل من هذه المواضيع تحقّق من [باقي مستندات React](/docs/hello-world.html). لتعلّم المزيد حول تعريف المكوّنات تحقّق من [مرجع واجهة برمجة التطبيق `React.Component`](/docs/react-component.html).
```
